### DISCUSSION TITLE
Flask 10.25 has been released! Look here for breaking changes.

### DISCUSSION THREAD

**Discussion by Montoya** [MetaMask/snaps#1198] [Announcements]
MetaMask Flask 10.25 has been released to Chrome and Firefox today. This includes many significant changes to the Snaps API which are detailed below. 

- Snaps are now required to request `endowment:rpc` to receive RPC requests with `onRpcRequest`
    - Snaps must specify if they want to receive RPC requests from dapps or snaps using `endowment:rpc: { dapps: true, snaps: true }` ([Documentation here](https://docs.metamask.io/guide/snaps-permissions.html#endowment-rpc))
- `wallet_enable` and `wallet_installSnaps` have been removed in favor of `wallet_requestSnaps` ([Documentation here](https://docs.metamask.io/guide/snaps-rpc-api.html#wallet-requestsnaps))
  - To connect to a snap from a dapp, use: `window.ethereum.request({ method: 'wallet_requestSnaps', params: { [snapId]: { version } }}`
- `wallet_requestSnaps` will now throw an error and roll back any changes made if installation of any snap fails. This means if you attempt to install multiple snaps in one request and one of those snaps fails to install, the entire set will be rolled back. 
- `snap_dialog` and transaction insights now both exclusively use custom UI ([Documentation here](https://docs.metamask.io/guide/snaps-concepts.html#custom-ui))
  - Please migrate all instances of `snap_confirm` to `snap_dialog`. `snap_confirm` is deprecated and will be removed. 
- `snap_manageState` now uses SIP-6 algorithm for encryption
  - This breaks all existing snaps that use `snap_manageState`; these snaps need to be reinstalled upon updating
- The `wallet` global used by snaps has been removed and replaced with two new globals: `snap` and `ethereum`
  - `ethereum` is an EIP-1193 provider and can be accessed by requesting the `endowment:ethereum-provider` permission
  - `snap` is always exposed and can be used to access snap specific functions using `snap.request()`
- Named parameters are required in all Snaps RPC methods
  - `snap_notify`
  - `snap_manageState`
  - `wallet_invokeSnap`
- `wallet_invokeSnap` does not support undefined parameters anymore
  - `window.ethereum.request({ method: 'wallet_invokeSnap', params: { snapId: 'foo', request: { method: 'bar', params: undefined } } })` is not valid, instead you must omit undefined `params`
- Removed `snap_getAppKey`
  - This has been replaced by `snap_getEntropy` ([Documentation here](https://docs.metamask.io/guide/snaps-rpc-api.html#snap-getentropy))
- The package `@metamask/snap-types` has been renamed to `@metamask/snaps-types`

Update: unfortunately the Cron Jobs API is broken in this release. A fix will be included in the next release of Flask. In the meantime, you can build Flask from the repository or install an older version if you need to use Cron Jobs.

**Comment by skunal8197** [comment]
As of this update,

After successfully connecting to my snap using `wallet_requestSnaps`,  
I am **unable** to successfully call `wallet_invokeSnap`. 

I am attempting to call it like so 
``` 
window.ethereum.request({
    method: 'wallet_invokeSnap',
    params: {
        snapId: snapId,
        request: {
            method: 'someMethod',
        },
    },
});
```

and am getting the following error

**`MetaMask - RPC Error: Expected params to be a single object. `**


Any assistance would be greatly appreciated!

**Reply by ziad-saab** [reply]
> As of this update,
> 
> After successfully connecting to my snap using `wallet_requestSnaps`, I am **unable** to successfully call `wallet_invokeSnap`.
> 
> I am attempting to call it like so
> 
> ```
> window.ethereum.request({
>     method: 'wallet_invokeSnap',
>     params: {
>         snapId: snapId,
>         request: {
>             method: 'someMethod',
>         },
>     },
> });
> ```
> 
> and am getting the following error
> 
> **`MetaMask - RPC Error: Expected params to be a single object. `**
> 
> Any assistance would be greatly appreciated!

Can you please double-check which version of MetaMask Flask you're using? Make sure you're not using a local build or something? The code you shared is the correct way to invoke a snap.

**Reply by skunal8197** [reply]
I am using MetaMask Flask Version `10.25.0-flask.0`. 

After a clean install of MetaMask Flask, I tried using two different versions of the `@metamask/snaps-cli` package in my snap and came across the following errors. 

When using ^0.18.1 to ^0.28.0 - I am able to connect and then when trying to call `wallet_invokeSnap` with this params object
```
params: {
        snapId: snapId,
        request: {
            method: 'someMethod',
        },
    },
```
I get this error - `MetaMask - RPC Error: Expected params to be a single object. `


When using ^0.29.0 - I am unable to connect and get this error.
`Failed to fetch Snap "local:http://localhost:8083/… manifest shasum does not match computed shasum..`

This is the code I am using to connect to the snap for reference.
```
     const snapId = `local:${window.location.href}`;
     async function connect() {
      await ethereum.request({
        method: 'wallet_requestSnaps',
        params: {
          [snapId]: {},
        },
      });
    }
```

**Reply by ziad-saab** [reply]
Regarding using version 0.29 of `mm-snap` it's normal, the extension is still not at that version for shasum calculations. Actually if you look at the [template-snap-monorepo](https://github.com/MetaMask/template-snap-monorepo) project, we're currently using version `0.27.1` of the Snaps CLI. I'm curious what would happen if you were to use that version. Can you check and report back please? Everything else looks good from what I can see. For reference, you can look at how we call `requestSnaps` here:

https://github.com/MetaMask/template-snap-monorepo/blob/main/packages/site/src/utils/snap.ts#L25

And how we call `invokeSnap` here:

https://github.com/MetaMask/template-snap-monorepo/blob/main/packages/site/src/utils/snap.ts#L58

Also wanted to double check, is it possible that you'd have two versions of the extension installed in your browser?

**Comment by skunal8197** [comment]
It finally worked! 

I'm not sure what exactly made it work..
It might have been my upgrade from `0.18.1` to `0.27.1` of Snaps CLI. 

The working code is as follows 
```
  async function connect() {
      await ethereum.request({
        method: 'wallet_requestSnaps',
        params: {
          [snapId]: {},
        },
      });
    }

    async function someMethod() {
      try {
        const response = await ethereum.request({
          method: 'wallet_invokeSnap',
          params: {
            snapId: snapId,
            request: { method: 'someMethod' },
          },
        });
        console.log(response);
        alert(JSON.stringify(response));
      } catch (err) {
        console.error(err);
        alert('Problem happened: ' + err.message || err);
      }
    }
```

Thanks so much for the help!

**Comment by yathishram** [comment]
Currently am facing issue with couple of things. I was able to add the `wallet_requestSnaps` to install snaps successfully. But when I try to create an RPC request to manage state or even create a transaction am facing the following issue.

```
{
    "code": -32603,
    "message": "wallet is not defined",
    "data": {
        "originalError": {}
    }
}
```

I see that wallet is not a global and it's changed. But haven't found anything on how to use it on docs. Can someone help me out with this?

**Reply by MumukshTayal** [reply]
You need to use @metamask/snaps-types instead of @metamask/snap-types. Notice the extra 's' in 'snaps-types'. Also, don't forget to change this while importing within your index.tsx file. They have mentioned this in the last point of the updates above. They have changed the global from 'wallet' to 'snap'. So you need to replace wallet.request() with snap.request()

**Comment by luketchang** [comment]
After the upgrade, when trying to connect/reconnect our snap, we get the following error `Failed to fetch Snap "local:http://localhost:8080": Invalid Snap manifest: manifest shasum does not match computed shasum..`. 

Versions 
```
"@metamask/snaps-cli": "^0.27.1",
"@metamask/snaps-types": "^0.27.1",
```

Code
```
export const connectSnap = async (
  params: Record<"version" | string, unknown> = {}
) => {
  await window.ethereum.request({
    method: "wallet_requestSnaps",
    params: {
      [SNAP_ID]: {
        ...params,
      },
    },
  });
};
```

Snap Manifest
```
{
  "version": "1.0.0-alpha",
  "description": "An example Snap written in TypeScript.",
  "proposedName": "TypeScript Example Snap",
  "repository": {
    "type": "git",
    "url": "https://github.com/MetaMask/template-snap-monorepo.git"
  },
  "source": {
    "shasum": "rHcB+GY7O3ym1Fi7MfiGgau8dQ6uaOCCEfwm1SyKsOo=",
    "location": {
      "npm": {
        "filePath": "dist/bundle.js",
        "iconPath": "images/icon.svg",
        "packageName": "@nocturne-xyz/snap",
        "registry": "https://registry.npmjs.org/"
      }
    }
  },
  "initialPermissions": {
    "endowment:rpc": {
      "dapps": true,
      "snaps": true
    },
    "snap_confirm": {},
    "endowment:network-access": {},
    "endowment:long-running": {},
    "snap_manageState": {},
    "snap_getBip44Entropy": [
      {
        "coinType": 6789
      }
    ]
  },
  "manifestVersion": "0.1"
}

```

**Reply by ghost** [reply]
same here

**Reply by FrederikBolding** [reply]
Hey there! Please update your `snaps-cli` to at least `^0.30.0` and rebuild your snap. That should create a new checksum that is accepted by Flask. Release notes for the new version of Flask are on the way!

**Reply by FrederikBolding** [reply]
Announcement here: https://github.com/MetaMask/snaps-monorepo/discussions/1348

**Reply by ritave** [reply]
@luketchang @enu-kuro @SotaTek-TungNguyen2 @Sotatek-HieuMai

**Reply by Sotatek-HieuMai** [reply]
thank you guys !

**Comment by 1337-ishaan** [comment]
In my case, it was due to the okx wallet installation, disable that.
